aten/src/ATen/native/cuda/TriangularOps.cu

```diff
@@ -19,33 +19,53 @@
 
 #include <ATen/cuda/CUDAApplyUtils.cuh>
 
+#define BOOL_SWITCH(COND, CONST_NAME, ...)      \
+  [&] {                                         \
+    if (COND) {                                 \
+      constexpr static bool CONST_NAME = true;  \
+      return __VA_ARGS__();                     \
+    } else {                                    \
+      constexpr static bool CONST_NAME = false; \
+      return __VA_ARGS__();                     \
+    }                                           \
+  }()
+
 namespace at::native {
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ triu/tril ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-template <typename scalar_t, typename IndexType, bool upper>
-C10_LAUNCH_BOUNDS_1(cuda::getApplyBlockSize())
+constexpr static int block_size = 128;
+
+template <typename scalar_t, typename IndexType, bool upper, int elements_per_thread, bool inplace>
```

Comment:
It looks like there is a good amount of additional specialization that we're adding here. Did you find that specializing on `inplace` made a large performance difference over just making this a runtime parameter?

(just curious as we want to balance absolute runtime perf against binary size)

Reply:
If "inplace" is used as a runtime parameter, the performance of `triu_` becomes 87%~95%, and the performance of `triu` becomes 98%~102%, with the exception of a few outliers..

Due to slight environmental differences (e.g., variations in background processes), I retested the performance of "inplace as template," and there may be a slight difference compared to the measurements taken three weeks ago.

  | function | dtype | shape | k | as template | as param | rel. perf
-- | -- | -- | -- | -- | -- | -- | --
various   dtype |   |   |   |   |   |  
  | triu_ | int8 | [1, 3072, 3072] | 0 | 0.028 | 0.028 | 0.98x
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.030 | 0.032 | 0.94x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.046 | 0.051 | 0.89x
  | triu_ | float64 | [1, 3072, 3072] | 0 | 0.083 | 0.088 | 0.95x
  | triu | int8 | [1, 3072, 3072] | 0 | 0.054 | 0.054 | 1.01x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.047 | 0.048 | 0.99x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.089 | 0.090 | 0.98x
  | triu | float64 | [1, 3072, 3072] | 0 | 0.173 | 0.173 | 1.00x
various   shape |   |   |   |   |   |  
  | triu_ | float32 | [1, 8192, 8192] | 0 | 0.309 | 0.346 | 0.89x
  | triu_ | float32 | [4, 1024, 1024] | 0 | 0.022 | 0.025 | 0.90x
  | triu_ | float32 | [4, 1021, 1021] | 0 | 0.023 | 0.026 | 0.90x
  | triu_ | float32 | [256, 128, 256] | 0 | 0.037 | 0.042 | 0.88x
  | triu_ | float32 | [128, 257, 125] | 0 | 0.029 | 0.033 | 0.87x
  | triu_ | float32 | [20480, 16, 16] | 0 | 0.036 | 0.041 | 0.87x
  | triu | float32 | [1, 8192, 8192] | 0 | 0.600 | 0.603 | 0.99x
  | triu | float32 | [4, 1024, 1024] | 0 | 0.042 | 0.042 | 1.01x
  | triu | float32 | [4, 1021, 1021] | 0 | 0.042 | 0.041 | 1.01x
  | triu | float32 | [256, 128, 256] | 0 | 0.093 | 0.093 | 1.00x
  | triu | float32 | [128, 257, 125] | 0 | 0.035 | 0.036 | 0.98x
  | triu | float32 | [20480, 16, 16] | 0 | 0.059 | 0.061 | 0.98x
various dim |   |   |   |   |   |  
  | triu_ | float32 | [3072, 3072] | 0 | 0.037 | 0.042 | 0.89x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.046 | 0.051 | 0.89x
  | triu_ | float32 | [1, 1, 3072, 3072] | 0 | 0.053 | 0.059 | 0.89x
  | triu | float32 | [3072, 3072] | 0 | 0.088 | 0.088 | 1.00x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.089 | 0.090 | 0.98x
  | triu | float32 | [1, 1, 3072, 3072] | 0 | 0.089 | 0.094 | 0.95x
various k |   |   |   |   |   |   |  
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.030 | 0.032 | 0.94x
  | triu_ | float16 | [1, 3072, 3072] | 1536 | 0.042 | 0.041 | 1.02x
  | triu_ | float16 | [1, 3072, 3072] | -1536 | 0.020 | 0.022 | 0.94x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.047 | 0.048 | 0.99x
  | triu | float16 | [1, 3072, 3072] | 1536 | 0.039 | 0.042 | 0.94x
  | triu | float16 | [1, 3072, 3072] | -1536 | 0.058 | 0.057 | 1.01x

Reply:
Although a template parameter, `elements_per_thread`, has been added, it corresponds one-to-one with `scalar_t`, so it does not increase the number of template specializations.